File: benchmarks/fixtures/webapp_go/internal/tasks/email_task.go
package tasks

import (
    "webapp_go/pkg/logger"
    "webapp_go/internal/services/email"
)

var emailTaskLog = logger.GetLogger("tasks.email")

// EmailTask represents a background email sending task.
type EmailTask struct {
    Sender     *email.EmailSender
    Recipients []string
    Subject    string
    Body       string
    Completed  bool
}

// NewEmailTask creates a new email task.
func NewEmailTask(sender *email.EmailSender, recipients []string, subject, body string) *EmailTask {
    emailTaskLog.Info("Creating EmailTask: subject=%s, recipients=%d", subject, len(recipients))
    return &EmailTask{
        Sender:     sender,
        Recipients: recipients,
        Subject:    subject,
        Body:       body,
        Completed:  false,
    }
}

// Execute runs the email task.
func (t *EmailTask) Execute() error {
    emailTaskLog.Info("Executing email task: %s", t.Subject)
    sent, err := t.Sender.SendBulk(t.Recipients, t.Subject, t.Body)
    if err != nil {
        emailTaskLog.Error("Email task failed: %v", err)
        return err
    }
    t.Completed = true
    emailTaskLog.Info("Email task completed: %d/%d sent", sent, len(t.Recipients))
    return nil
}

// Status returns the task status.
func (t *EmailTask) Status() string {
    if t.Completed {
        return "completed"
    }
    return "pending"
}

File: benchmarks/fixtures/webapp_rb/tasks/email_task.rb
# Email background tasks.

require_relative '../utils/helpers'
require_relative '../database/connection'
require_relative '../services/email/sender'

# Send welcome email.
def send_welcome_email(user_data, db)
  get_logger('tasks.email').info("Sending welcome email to #{user_data[:email]}")
  sender = EmailSender.new(db)
  sender.send_template(user_data[:email], 'welcome', { 'name' => user_data[:name] })
end

# Send password reset email.
def send_password_reset_email(email, reset_link, db)
  get_logger('tasks.email').info("Sending password reset to #{email}")
  sender = EmailSender.new(db)
  sender.send_template(email, 'password_reset', { 'link' => reset_link })
end

# Process email queue.
def process_email_queue(db)
  get_logger('tasks.email').info('Processing email queue')
  sender = EmailSender.new(db)
  pending = db.find_all('notifications', { channel: 'email', status: 'pending' })
  sent = 0
  failed = 0
  pending.each do |n|
    begin
      sender.send_email(n[:user_id], n[:subject], n[:body])
      sent += 1
    rescue StandardError
      failed += 1
    end
  end
  { sent: sent, failed: failed }
end
